wallet-new/src/Cardano/Wallet/Kernel/Wallets.hs

```diff
@@ -142,3 +184,81 @@ createWalletHdRnd pw hasSpendingPassword name assuranceLevel esk utxo = do
+               -> HD.AssuranceLevel
+               -> HD.WalletName
+               -> IO (Either HD.UnknownHdRoot (Kernel.DB, HdRoot))
+updateHdWallet pw hdRootId assuranceLevel walletName = do
+    res <- update' (pw ^. wallets) (UpdateHdWallet hdRootId assuranceLevel walletName)
+    case res of
+         Left e              -> return (Left e)
+         Right (db, newRoot) -> return (Right (db, newRoot))
+
+updatePassword :: PassiveWallet
+               -> HD.HdRootId
+               -> PassPhrase
+               -- ^ The old 'PassPhrase' for this Wallet.
+               -> PassPhrase
+               -- ^ The new 'PassPhrase' for this Wallet.
+               -> IO (Either UpdateWalletPasswordError (Kernel.DB, HdRoot))
+updatePassword pw hdRootId oldPassword newPassword = do
+    let keystore = pw ^. walletKeystore
+        wId = WalletIdHdRnd hdRootId
+    -- STEP 1: Lookup the key from the keystore
+    mbKey <- Keystore.lookup wId keystore
+    case mbKey of
+         Nothing -> return $ Left $ UpdateWalletPasswordKeyNotFound hdRootId
+         Just esk -> do
+             -- STEP 2: Check that the 2 password matches. While this in
+             --         principle could be checked on the wallet layer side,
+             --         it's preferrable to do everything in the kernel to make
+             --         this operation really atomic.
+             let pwdCheck = hoistMaybeWith (UpdateWalletPasswordOldPasswordMismatch hdRootId) $
+                            checkPassMatches oldPassword esk
+
+             -- STEP 3: Compute the new key using the cryptographic primitives
+             --         we have. This operation doesn't change any state, it
+             --         just recomputes the new 'EncryptedSecretKey' in-place.
+             genNewKey <- changeEncPassphrase oldPassword newPassword esk
+             let mbNewKey = hoistMaybeWith (UpdateWalletPasswordChangeFailed hdRootId) $
+                            genNewKey
+
+             case pwdCheck >> mbNewKey of
+                  Left e -> return (Left e)
+                  Right newKey -> do
+                      -- STEP 4: Update the keystore, atomically.
+                      Keystore.replace wId newKey keystore
+                      -- STEP 5: Update the timestamp in the wallet data storage.
+                      lastUpdateNow <- InDb <$> getCurrentTimestamp
+                      let hasSpendingPassword = HD.HasSpendingPassword lastUpdateNow
+                      res <- update' (pw ^. wallets)
+                                     (UpdateHdRootPassword hdRootId hasSpendingPassword)
+                      case res of
+                           Left e ->
+                               return $ Left (UpdateWalletPasswordUnknownHdRoot e)
+                           Right (db, hdRoot') -> return $ Right (db, hdRoot')
+
+-- | Hoist a Maybe into an Either so that multiple calls that may fail can
+-- be chained.
+hoistMaybeWith :: e -> Maybe a -> Either e a
```

Comment:
`universum` provides `maybeToRight` for this purpose.

Reply:
Ah!